src/view.c

```diff
@@ -18,6 +18,66 @@
 #define LAB_FALLBACK_WIDTH  640
 #define LAB_FALLBACK_HEIGHT 480
 
+static bool
+matches_criteria(struct view *view, enum lab_view_criteria criteria)
+{
+	if (!isfocusable(view)) {
+		return false;
+	}
+	if (criteria & LAB_VIEW_CRITERIA_CURRENT_WORKSPACE) {
+		/*
+		 * Always-on-top views are always on the current desktop and are
+		 * special in that they live in a different tree.
+		 */
+		struct server *server = view->server;
+		if (view->scene_tree->node.parent != server->workspace_current->tree
+				&& !view_is_always_on_top(view)) {
+			return false;
+		}
+	}
+	if (criteria & LAB_VIEW_CRITERIA_NO_ALWAYS_ON_TOP) {
+		if (view_is_always_on_top(view)) {
+			return false;
+		}
+	}
+	if (criteria & LAB_VIEW_CRITERIA_NO_SKIP_WINDOW_SWITCHER) {
+		if (window_rules_get_property(view, "skipWindowSwitcher") == LAB_PROP_TRUE) {
+			return false;
+		}
+	}
+	return true;
+}
+
+struct view *
+view_next(struct wl_list *head, struct view *view, enum lab_view_criteria criteria)
+{
+	assert(head);
+
+	struct wl_list *ref_item = view ? &view->link : head;
+	for (;;) {
+		ref_item = ref_item->next;
+		if (ref_item == head) {
+			return NULL;
+		}
+		view = wl_container_of(ref_item, view, link);
+		if (matches_criteria(view, criteria)) {
+			return view;
+		}
+	}
+	return NULL; /* Should never be reached */
+}
+
+void
+view_array_append(struct server *server, struct wl_array *views,
+		enum lab_view_criteria criteria)
+{
+	struct view *view;
+	for_each_view(view, &server->views, criteria) {
+		struct view **entry = wl_array_add(views, sizeof(*entry));
```

Comment:
wl_array_add can fail. Dunno if we wanna recover or not, but explicit assert IMO would be better.

Reply:
Done